src/lib.rs

```diff
@@ -15,18 +15,51 @@
 //! assert_eq!(now.elapsed().as_secs(), 9);
 //! ```
 use std::{
+    debug_assert,
     io::{self, Read, Write},
-    time::Duration,
+    time::{Duration, Instant},
 };
 
+const ACCEPTABLE_SPEED_DIFF: f64 = 0.1 / 100.0;
+
+#[derive(Clone, Debug)]
 pub struct LimiterOptions {
-    window_length: u128,
-    window_time: Duration,
-    bucket_size: usize,
+    pub window_length: u64,
+    pub window_time: Duration,
+    pub bucket_size: u64,
 }
 
 impl LimiterOptions {
-    pub fn new(window_length: u128, window_time: Duration, bucket_size: usize) -> LimiterOptions {
+    pub fn new(mut window_length: u64, mut window_time: Duration, mut bucket_size: u64) -> LimiterOptions {
+        let rate = window_length.min(bucket_size) as f64;
+        let tw: f64 = window_time.as_nanos() as f64;
+        let init_speed = (rate / tw) * 1_000_000.0;
+        // println!("Opts: {} {:?} {} ({} b/s)", window_length, window_time, bucket_size, init_speed);
+        
+        let mut new_speed = init_speed;
+        let mut new_wlen = window_length;
+        let mut new_wtime = window_time;
+        let mut new_bsize = bucket_size;
+        while ((new_speed / init_speed) - 1.0).abs() < ACCEPTABLE_SPEED_DIFF {
```

Comment:
Can you place a comment at the top of this block of code because it's not trivial to understand and i'm having hard time to figure out what's going on

Reply:
Added comments, hope it's better now :sweat_smile: